### DISCUSSION TITLE
Using the 'Mask' Tool

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#747] [General]
I am wishing to even out the UV light intensity across a printer screen. 

I am aware of Jan Mrazek's approach for this, but the github/Marlin/automation side of it is beyond my skills.

I came across [Devoney/CreateMask](https://github.com/Devoney/CreateMask) and this feels more straight forward, but upon testing a simple 3x7 grid on a Mars 3, I see that the software is creating a 3x7 pixel bitmap of a mask and resizing that to the 4098x2560 resolution of a Mars 3. I think this method will work, but I suspect I just need way more data points.

What I'm wondering is are there any other common approaches to creating a mask? Is this something that could possible be implemented within UVtools itself?

**Comment by sn4k3** [comment]
Creating a good mask always need manual work, it's boring and somehow hard for most. The more points you measure more precise it can be. In past I saw devices made for that but the different LCD sizes makes hard to have 1 universal product.
Other than applying the mask to layers, UVtools can't help further.

**Reply by aWildTomAppeared** [reply]
I made a mask for my Saturn 1 Here are the results
![image](https://github.com/sn4k3/UVtools/assets/148269455/3673118a-367b-4d68-b03c-7f7a64fb2ead)

I made the mask just by putting my printer sideways, aligning a camera about 15ft away then taking photos while the printer was exposing then doing some layer adjustments to get a mask. Originally I adjusted the mask to work with subtraction as that is what the UV tools mask applier does, except it doesn't really, your tool gives results that look exactly like the AbsDif import type, due to this being an absolute operation it's not smooth at all, very unusable. So then I tried importing the mask with a subtract layer type and this looked much better however subtracting just isn't the right operation for this, it's correct for 100% layers but with lower values it takes away too much. The correct operation is multiply. Using the blender compositor I applied my mask with a multiply operation and the results were perfect (although blender compositor takes seconds to output each layer so this isn't usable, I'd love to do it in UVT). 

![image](https://github.com/sn4k3/UVtools/assets/148269455/37845f0d-ae48-48c9-be39-dc10b0788142)


in that image, the middle column is multiply masked images of the printer with levels of pixel brightness from 10%-20%...90%-100%, and on the left it's the same images but with the exposure compensated so they look about the same brightness so you can see the difference better. On the right is an exposure with a subtract operation, I can't remember what brightness that was at, probably 20%, but as you can see because it's just subtracting at these lower exposures it takes away too much from the pixels, the correct operation is multiplied. So please add a multiply layer import type because that would be useful for this and more, and if you do that just set the mask operation to use it and it will be perfect....

Well not exactly, masking isn't great because most printers have issues when you use a lot of grey values, the layers jump about it's a mess. Anycubic printer files have their grey values limited to 16 I would very much like to try to put this up as I have made a vertical AA slicer and 16 levels isn't quite enough, but for the Elegoos I'd like to put it down to say 64 or something to see if that helps fix the issue of images glitching about when there are too many grey values. But while I can see such print parameters, I can't edit them in UV tools, is there a way and I'm missing it? If not then please add the ability to edit the properties in the left tab.  I have tried exporting files to folders, changing to .ini files there, and then compressing to a .zip or .rar and opening that but then I just get an error in UVT.

**Reply by sn4k3** [reply]
Currently the mask tool uses the "Bitwise And" operator which is different from Sub and AbsDiff.

The tool pixel arithmetic gives many methods which can be used to apply masks, multiply included, and let you to load an image as pattern to apply. However, I don't see the multiply operator to work as you want in UVtools, because it works with 0-255 values. If you have a white pixel of "255" then multiply by 0 results in 0 and other value than 0 will result in "255", you can't fade pixels using this method as it is because the lack of float. As I understand you are using multiply with float values, right? If that the case this can be obtain via scripting where you can convert layer to float multiply and then back to bytes.

> But while I can see such print parameters, I can't edit them in UV tools, is there a way and I'm missing it? 

Which parameters are you talking about? All you can edit are on "Edit print parameters"

**Reply by aWildTomAppeared** [reply]
so to make a mask for multiplication is different than subtraction. Here is my Saturn 1 mask for multiplication. It's fairly light because well if it had values of 0 that would mean it would turn a white pixel into black and I don't want do darken it that much, the darkest area is about 30% which means over a white area it will multiply 1x0.7 to give 0.7. This is how the multiply layer type works in photoshop or anything. yes 255x255 wouldn't work, it needs to be converted to a 0-1 float. I'll see if I can script it if that's the only option but multiplication is how masking should be done, subtraction is bad and bitwise functions are awful for this so it would be nice to have in the tool in there.
![Saturn 1 Mask 4b](https://github.com/sn4k3/UVtools/assets/148269455/19422ed2-15ba-45f7-a7ae-24843c1edfa7)


the properties I'm talking about are these

![image](https://github.com/sn4k3/UVtools/assets/148269455/5f26459b-be03-421b-9072-2c37152f5355)


is it not possible to change these?

also side note, two bugs I've found in 4.0.2, one is when importing layers with the same size as the file even when using modes like replace, insert, subtract etc it always says the layer is too big when it isn't, it's just fine on 3.15.1 (the other version I have) but not on 4.0.2.  Also, much less minor but when you don't type anything into the move tool, leave a field blank, instead of the program filling it with a 0 as it used to, now it shows an error.

**Reply by sn4k3** [reply]
If you full break-down the procedure step by step I can look into it, I do not fully understand how the operation you doing it's made. For example if you feed and 8 bit image and then it's converted to 0f-1f to do the mul or anything else. If you want try the script approach and after obtain the goal I can also incorporate it.

> is it not possible to change these?

No, that table is informative only and the raw fields of the format, most have zero impact on the print, they are just infos/dumps. All you can edit is on Tools - Edit print parameters which will update and reflect on that table. However, if you need you can edit everything using File - Terminal.
Which of the field's you want to edit?

> one is when importing layers with the same size as the file even when using modes like replace, insert, subtract etc it always says the layer is too big when it isn't, it's just fine on 3.15.1 (the other version I have) but not on 4.0.2.

Provide me the file samples so I can debug.


> Also, much less minor but when you don't type anything into the move tool, leave a field blank, instead of the program filling it with a 0 as it used to, now it shows an error.

This is new to me, looks like something from UI framework is different now. I will probe, but every numeric input have that problem on UVtools. I wasn't aware.

**Reply by aWildTomAppeared** [reply]
Okay so I looked it up and CvInvoke has a multiply modifier https://emgu.com/wiki/files/3.0.0/document/html/f7e066ef-e043-669f-c3ba-74b13355c8cb.htm

it has a scale parameter in which you can just put in 1/256 aka about 0.0039 

                        case ImportTypes.Multiply:
                            {
                                if (layerIndex >= SlicerFile.LayerCount) return;
                                using var originalMat = SlicerFile[layerIndex].LayerMat;
                                using var newMat = fileFormat[i].LayerMat;
                                CvInvoke.Multiply(originalMat, newMat, newMat, 0.0039);  //0.00390625 would be more accurate
                                SlicerFile[layerIndex].LayerMat = newMat;
                                break;
                            }

I tested this and it works here is the result of a multiplied layer (I put this into the layer import thing file because while it should be the default for the mask editor it would also be nice to have a multiply layer import type)
![image](https://github.com/sn4k3/UVtools/assets/148269455/6646be90-cace-498e-a3ea-107eac45888b)

because the layer was full brightness it does 1x the mask so it looks exactly like the mask where the model is, but if it was darker before it would be say 0.5x the mask or rather 256 x the mask / (1/256)

I had to do this in 3.15.1 though because as I mentioned many import types just don't work 
![image](https://github.com/sn4k3/UVtools/assets/148269455/7ac6fc74-90e7-4636-a1fe-9765b96c83de)
I get this error and if I hide the error well it doesn't appear but the operations still don't work so idk what's up with that. 
It happens with every file I have, it would do the same for you? it even happens if I scale up the file and import a single image so that I know the image is smaller and less layer than the file. 

This .zip has a ctb file from me that this happens with and my mask image and the multiply modified "OperationLayerImport.cs" file for 3.15.15, I just added the method-397, 'a blank check if is possible to process this file'-369 and the desctiption-31:

[UVT Debug Files.zip](https://github.com/sn4k3/UVtools/files/13291309/UVT.Debug.Files.zip)

**Comment by aWildTomAppeared** [comment]
Okay so I just fixed the mask, it's literally just one line change 
![image](https://github.com/sn4k3/UVtools/assets/148269455/73f6ea6e-64e1-4bce-a3cd-ba76266e278b)


just replace the current line with this:   CvInvoke.Multiply(target, Mask, target, 0.00390625);
Here are the results:
![image](https://github.com/sn4k3/UVtools/assets/148269455/71e302a5-dbc8-4c01-a207-1c8d8fa95cae)

**Reply by sn4k3** [reply]
Will add that to next patch

**Reply by aWildTomAppeared** [reply]
Nice, Another thing very similar, in line 730 of OperationPixelArithmetic.cs there is the pixel multiply operator 

![image](https://github.com/sn4k3/UVtools/assets/148269455/0ad094ba-d38c-4ab8-8730-44df888d9ba3)

CvInvoke.Multiply(target, tempMat, target, 0.003900625);

this should also use the 0.003900625 scale because otherwise multiply here is useless, like how can you do anything to an image if you are multiplying by 25g, it's always going to be massive, the only use for multiply is using numbers under 0 like 0.8 x 0.5 = 0.4, now we can't do decimals but we can use this scale so that you could instead do something like 200/256 x 128/256 = 100. 

Divide on the other hand could be kept the same because that makes some sense eg you can divide by 2 or 4 etc, if it was scaled then you could only divide by numbers under 1 which makes them bigger which isn't useful because you cant go above full brightness. 

I have some other ideas but I'll make new posts for them.
